modules/platforms/dotnet/Apache.Ignite/Internal/Table/Table.cs

```diff
@@ -67,106 +67,377 @@ public Table(string name, Guid id, ClientFailoverSocket socket)
             }
+
+            using var resBuf = await _socket.DoOutInOpAsync(ClientOp.TupleDeleteAll, writer).ConfigureAwait(false);
+            var resSchema = await ReadSchemaAsync(resBuf, schema).ConfigureAwait(false);
+
+            return ReadTuples(resBuf, resSchema, keyOnly: true);
         }
 
         /// <inheritdoc/>
-        public async Task UpsertAsync(IIgniteTuple rec)
+        public async Task<IList<IIgniteTuple>> DeleteAllExactAsync(IEnumerable<IIgniteTuple> records)
         {
-            IgniteArgumentCheck.NotNull(rec, nameof(rec));
+            IgniteArgumentCheck.NotNull(records, nameof(records));
 
             var schema = await GetLatestSchemaAsync().ConfigureAwait(false);
 
             using var writer = new PooledArrayBufferWriter();
-            Write(writer.GetMessageWriter());
+            var count = WriteTuples(writer, schema, records);
 
-            using var resBuf = await _socket.DoOutInOpAsync(ClientOp.TupleUpsert, writer).ConfigureAwait(false);
+            if (count == 0)
+            {
+                return Array.Empty<IIgniteTuple>();
+            }
+
+            using var resBuf = await _socket.DoOutInOpAsync(ClientOp.TupleDeleteAllExact, writer).ConfigureAwait(false);
+            var resSchema = await ReadSchemaAsync(resBuf, schema).ConfigureAwait(false);
+
+            return ReadTuples(resBuf, resSchema);
+        }
 
-            void Write(MessagePackWriter w)
+        private static IIgniteTuple? ReadValueTuple(PooledBuffer buf, Schema? schema, IIgniteTuple key)
+        {
+            if (schema == null)
             {
-                w.Write(Id);
-                w.Write(schema.Version);
+                return null;
+            }
 
-                foreach (var col in schema.Columns)
+            // Skip schema version.
+            var r = buf.GetReader();
+            r.Skip();
+
+            var columns = schema.Columns;
+            var tuple = new IgniteTuple(columns.Count);
+
+            for (var i = 0; i < columns.Count; i++)
+            {
+                var column = columns[i];
+
+                if (i < schema.KeyColumnCount)
+                {
+                    tuple[column.Name] = key[column.Name];
+                }
+                else
                 {
-                    var colIdx = rec.GetOrdinal(col.Name);
-
-                    if (colIdx < 0)
-                    {
-                        w.WriteNil();
-                    }
-                    else
-                    {
-                        w.WriteObject(rec[colIdx]);
-                    }
+                    tuple[column.Name] = r.ReadObject(column.Type);
                 }
+            }
 
-                w.Flush();
+            return tuple;
+        }
+
+        private static IIgniteTuple ReadTuple(ref MessagePackReader r, Schema schema, bool keyOnly = false)
+        {
+            var columns = schema.Columns;
+            var count = keyOnly ? schema.KeyColumnCount : columns.Count;
```

Comment:
Do key columns always go first?

Reply:
Yes, key columns always come first in the schema.